### PR TITLE
Parent interview report

### PR DESCRIPTION
**Pull request by aman-tunerlabs**
{{url}}/assessment/api/v1/reports/getParentInterviewResponses/PROGID01?fromDate=01-03-2019&toDate=02-03-2019

### CODE REVIEW

File: controllers/v1/reportsController.js [hunk 1]
```diff
@@ -2043,6 +2043,229 @@ module.exports = class Reports {
...
+        fetchRequiredSubmissionDocumentIdQueryObj["parentInterviewResponsesFieldArray.completedAt"]["$gte"] = fromDate
+        fetchRequiredSubmissionDocumentIdQueryObj["parentInterviewResponsesFieldArray.completedAt"]["$lte"] = toDate
+
+        const submissionDocumentIdsToProcess = await database.models.submissions.find(
+          fetchRequiredSubmissionDocumentIdQueryObj,
+          { _id: 1 }
+        ).lean()
+
+        let fileName = `ParentInterviewResponsesReport `;
+        (fromDate) ? fileName += "from date _" + fromDate : "";
```
Comment (ajithtunerlabs) [inline controllers/v1/reportsController.js]: add format for fromDate toDate

File: controllers/v1/reportsController.js [hunk 2]
```diff
@@ -2043,6 +2043,229 @@ module.exports = class Reports {
...
+            });
+
+            submissionDocuments = await database.models.submissions.find({
+              _id: { $in: submissionIds }
+            }, {
+                "schoolInformation.name": 1,
+                "schoolInformation.externalId": 1,
+                "schoolInformation.schoolTypes": 1,
+                "schoolInformation.administration": 1,
+                "parentInterviewResponsesFieldArray": 1
```
Comment (ajithtunerlabs) [inline controllers/v1/reportsController.js]: please project exactly what you want!

File: controllers/v1/reportsController.js [hunk 3]
```diff
@@ -2043,6 +2043,229 @@ module.exports = class Reports {
...
+        let toDate = req.query.toDate ? new Date(req.query.toDate.split("-").reverse().join("-")) : new Date()
+        toDate.setHours(23, 59, 59)
+
+        if (fromDate > toDate) {
+          throw "From date cannot be greater than to date."
+        }
+
+        let fetchRequiredSubmissionDocumentIdQueryObj = {};
+        fetchRequiredSubmissionDocumentIdQueryObj["programExternalId"] = req.params._id
+        fetchRequiredSubmissionDocumentIdQueryObj["parentInterviewResponses"] = { $exists: true }
```
Comment (ajithtunerlabs) [inline controllers/v1/reportsController.js]: change parentInterviewResponsesFieldArray -> parentInterviewResponsesArray

we will change these afterward.

File: controllers/v1/reportsController.js [hunk 4]
```diff
@@ -2043,6 +2043,229 @@ module.exports = class Reports {
...
+        fetchRequiredSubmissionDocumentIdQueryObj["parentInterviewResponsesFieldArray.completedAt"] = {}
+        fetchRequiredSubmissionDocumentIdQueryObj["parentInterviewResponsesFieldArray.completedAt"]["$gte"] = fromDate
+        fetchRequiredSubmissionDocumentIdQueryObj["parentInterviewResponsesFieldArray.completedAt"]["$lte"] = toDate
+
+        let submissionDocumentIds = await database.models.submissions.find(fetchRequiredSubmissionDocumentIdQueryObj, { _id: 1 }).lean()
+
+        let fileName = `ParentInterviewCallResponseReport`;
+        (fromDate) ? fileName += "from date _" + fromDate : "";
```
Comment (ajithtunerlabs) [inline controllers/v1/reportsController.js]: format the fromDate and toDate

File: controllers/v1/reportsController.js [hunk 5]
```diff
@@ -2043,6 +2043,229 @@ module.exports = class Reports {
...
+              let countForP5 = 0;
+              let countForP6 = 0;
+
+              result["schoolId"] = eachSubmission.schoolInformation.externalId;
+              result["schoolName"] = eachSubmission.schoolInformation.name;
+              result["School (SDMC, EDMC, DOE, NDMC, North DMC, DCB, Private)"] = eachSubmission.schoolInformation.administration;
+
+              eachSubmission.parentInterviewResponsesFieldArray.forEach(eachParentInterviewResponse => {
+                if ((eachParentInterviewResponse.completedAt >= fromDate && eachParentInterviewResponse.completedAt < toDate)) {
+                  eachParentInterviewResponse.parentInformation.type.forEach(eachParentType => {
```
Comment (ajithtunerlabs) [inline controllers/v1/reportsController.js]: make countForP* as generic.

File: controllers/v1/reportsController.js [hunk 6]
```diff
@@ -2043,6 +2043,229 @@ module.exports = class Reports {
...
+              let countForP1 = 0;
+              let countForP2 = 0;
+              let countForP3 = 0;
+              let countForP4 = 0;
+              let countForP5 = 0;
+              let countForP6 = 0;
+
+              result["schoolId"] = eachSubmission.schoolInformation.externalId;
+              result["schoolName"] = eachSubmission.schoolInformation.name;
+              result["School (SDMC, EDMC, DOE, NDMC, North DMC, DCB, Private)"] = eachSubmission.schoolInformation.administration;
```
Comment (ajithtunerlabs) [inline controllers/v1/reportsController.js]: unwind parent interview submissions take specific submissions don't load all submissions.

File: controllers/v1/reportsController.js [hunk 7]
```diff
@@ -2043,6 +2043,229 @@ module.exports = class Reports {
...
+            submissionIds = chunkOfSubmissionDocumentIds[pointerToSubmissionIdChunkArray].map(submissionModel => {
+              return submissionModel._id
+            });
+
+            submissionDocuments = await database.models.submissions.find({
+              _id: { $in: submissionIds }
+            }, {
+                "schoolInformation.name": 1,
+                "schoolInformation.externalId": 1,
+                "parentInterviewResponsesFieldArray.parentInformation": 1
```
Comment (ajithtunerlabs) [inline controllers/v1/reportsController.js]: unwind parent interview submissions take specific submissions don't load all submissions.

File: controllers/v1/reportsController.js [hunk 8]
```diff
@@ -2043,6 +2043,229 @@ module.exports = class Reports {
...
+              }
+            ).lean()
+
+            await Promise.all(submissionDocuments.map(async (eachSubmission) => {
+              eachSubmission.parentInterviewResponsesFieldArray.forEach(eachParentInterviewResponse => {
+                if (eachParentInterviewResponse.parentInformation.callResponse && eachParentInterviewResponse.parentInformation.callResponse == "R2") {
```
Comment (ajithtunerlabs) [inline controllers/v1/reportsController.js]: unwind parent interview submissions take specific submissions which are having callResponse as R2.

File: controllers/v1/reportsController.js [hunk 9]
```diff
@@ -2043,6 +2043,229 @@ module.exports = class Reports {
...
+    })
+  }
+
+  async parentInterviewCallResponses(req) {
+    return new Promise(async (resolve, reject) => {
+      try {
+        if (!req.query.fromDate) {
+          throw "From Date is mandatory"
+        }
+        let fromDate = new Date(req.query.fromDate.split("-").reverse().join("-"))
```
Comment (ajithtunerlabs) [inline controllers/v1/reportsController.js]: please make fromDate and toDate check as a generic one.  put it inside utils.

File: controllers/v1/reportsController.js (hunk 3 above)
Comment (aks30) [reply controllers/v1/reportsController.js]: Resolved.

File: controllers/v1/reportsController.js [hunk 10]
```diff
@@ -2155,6 +2156,352 @@ module.exports = class Reports {
...
+        return reject({
+          status: 500,
+          message: "Oops! Something went wrong!",
+          errorObject: error
+        });
+      }
+    })
+  }
+
+  async parentInterviewCall(req) {
```
Comment (ajithtunerlabs) [inline controllers/v1/reportsController.js]: Please provide a meaningful name

File: controllers/v1/reportsController.js [hunk 11]
```diff
@@ -2155,6 +2156,352 @@ module.exports = class Reports {
...
+        return reject({
+          status: 500,
+          message: "Oops! Something went wrong!",
+          errorObject: error
+        });
+      }
+    })
+  }
+
+  async parentInterviewCallResponses(req) {
```
Comment (ajithtunerlabs) [inline controllers/v1/reportsController.js]: Please provide a meaningful name

File: controllers/v1/reportsController.js [hunk 12]
```diff
@@ -2155,6 +2156,352 @@ module.exports = class Reports {
     })
   }
 
+
+  async getParentInterviewResponses(req) {
```
Comment (ajithtunerlabs) [inline controllers/v1/reportsController.js]: Please provide a meaningful name

**Pull request merged**